src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc

```diff
@@ -0,0 +1,774 @@
+          (result->header_fields.find("variable_header_length") == result->header_fields.end())) {
+        return ParseState::kInvalid;
+      }
+
+      result->payload["topic_filter"] = "";
+      payload_length = result->header_fields["remaining_length"] -
+                       result->header_fields["variable_header_length"];
+      while (payload_length > 0) {
+        PX_ASSIGN_OR_RETURN_INVALID(topic_filter_length, decoder->ExtractBEInt<uint16_t>());
+        PX_ASSIGN_OR_RETURN_INVALID(std::string_view topic_filter,
+                                    decoder->ExtractString(topic_filter_length));
+        if (result->payload["topic_filter"].empty()) {
+          result->payload["topic_filter"] += std::string(topic_filter);
+        } else {
+          result->payload["topic_filter"] += ", " + std::string(topic_filter);
+        }
+        payload_length -= (2 + topic_filter_length);
+      }
+      return ParseState::kSuccess;
+    }
+    case MqttControlPacketType::SUBACK:
+    case MqttControlPacketType::UNSUBACK: {
+      size_t payload_length;
+      uint8_t reason_code;
+
+      if ((result->header_fields.find("remaining_length") == result->header_fields.end()) ||
+          (result->header_fields.find("variable_header_length") == result->header_fields.end())) {
+        return ParseState::kInvalid;
+      }
+
+      result->payload["reason_code"] = "";
+      payload_length = result->header_fields["remaining_length"] -
+                       result->header_fields["variable_header_length"];
+      while (payload_length > 0) {
+        PX_ASSIGN_OR_RETURN_INVALID(reason_code, decoder->ExtractBEInt<uint8_t>());
+        if (result->payload["reason_code"].empty()) {
+          result->payload["reason_code"] += std::to_string(reason_code);
+        } else {
+          result->payload["reason_code"] += ", " + std::to_string(reason_code);
+        }
+        payload_length -= 1;
+      }
+      return ParseState::kSuccess;
+    }
+    case MqttControlPacketType::PINGREQ:
+    case MqttControlPacketType::PINGRESP:
+    case MqttControlPacketType::DISCONNECT:
+      return ParseState::kSuccess;
+    default:
+      return ParseState::kInvalid;
+  }
+}
+
+ParseState ParseFrame(message_type_t type, std::string_view* buf, Message* result) {
+  CTX_DCHECK(type == message_type_t::kRequest || type == message_type_t::kResponse);
+  if (buf->size() < 2) {
+    return ParseState::kNeedsMoreData;
+  }
+
+  BinaryDecoder decoder(*buf);
+
+  // Parsing the fixed header
+  // Control Packet Type extracted from first four bits of the first byte
+  PX_ASSIGN_OR_RETURN_NEEDS_MORE_DATA(uint8_t control_packet_code_flags,
+                                      decoder.ExtractBEInt<uint8_t>());
+  uint8_t control_packet_code = control_packet_code_flags >> 4;
+  uint8_t control_packet_flags = control_packet_code_flags & 0x0F;
+
+  MqttControlPacketType control_packet_type = GetControlPacketType(control_packet_code);
```

Comment:
`GetControlPacketType` can be replaced with `magic_enum::enum_cast<MqttControlPacketType>`. If the resulting `std::optional<MqttControlPacketType>` doesn't have a value that should be equivalent to the `MqttControlPacketType::INVALID` type. It seems like we would want to return `kInvalid` for that case anyway?

Reply:
I have replaced this with magic_enum.